### PR TITLE
Add error suggestion per #1023

### PR DESCRIPTION
**Pull request by mjgpy3**
See #1023, not sure if that specific case spans across other nearby errors (e.g. `KindsDoNotUnify`, `ConstrainedTypeUnified`).

Any feedback is much appreciated.

### CODE REVIEW

**Comment by paf31** [comment]
I think this is good, but will apply in too many cases. We don't want _every_ unification error to give this message, just those from misuse of the `.` syntax.

**Comment by mjgpy3** [comment]
Gotcha, I'll keep looking into it

**Comment by mjgpy3** [comment]
@paf31 I limited the cases slightly, still not entirely sure if it's right. The issue spells out a case where it thinks it's looking at a `Prim.Object` and a `Prim.Function`, is that the only case? Can you provide any advice on how to make this more specific?

**Comment by mjgpy3** [comment]
Looks like tests took too long...

**Comment by garyb** [comment]
I think Travis is having issues at the moment generally, but I've restarted the build.

**Comment by paf31** [comment]
I think we probably do want to restrict this to `Prim.Object` and `Prim.Function`. Also, I liked the previous approach of handling this in `Errors`, since eventually I'd like all "clean up" passes to go there. The typechecker shouldn't really be generating English text directly any more.

**Comment by mjgpy3** [comment]
Gotcha, thanks! Can you give me an insight into how to restrict to those? I started looking at `TypeConstructor` but I ended up with very ugly pattern match that reached deeply into `ProperName`.

**Comment by paf31** [comment]
I think you can just test `== tyObject` and `== tyFunction`, after looking under `TypeConstructor`.

**Comment by mjgpy3** [comment]
Latest push should be close

**Comment by paf31** [comment]
Sorry to drag this out, but I don't really want to add even more constructors to `ErrorMessage` to deal with special cases. This sort of thing belongs in the error printer. As I hinted above: "I liked the previous approach of handling this in Errors, since eventually I'd like all "clean up" passes to go there." Matching particular classes of errors is already tricky.

**Comment by mjgpy3** [comment]
Alright, will update, thanks

**Comment by mjgpy3** [comment]
Sorry about all the back and forth on this, I hope the current push is reasonable :)

**Comment by paf31** [comment]
No problem at all. I just want to make sure we get this right since I'd like to add more things like this. Thanks for the help!

**Comment by mjgpy3** [comment]
@paf31 this isn't done, just curious if it matches what you're thinking -- if it looks okay, I'll handle the other wrapped-error types

**Comment by paf31** [comment]
I think it's definitely on the right track :+1:

**Comment by mjgpy3** [comment]
Is there a more concise way to unwrap those? I saw `simplifyErrorMessage` but I don't think it does exactly what I want.

**Comment by paf31** [comment]
I don't think so, but it might be worth extracting an `unwrapErrorMessage` function or something like that, to pull off the outer data constructors.

**Comment by mjgpy3** [comment]
Alright, will do. Thanks again!

**Comment by mjgpy3** [comment]
Any feedback is greatly appreciated.

File: src/Language/PureScript/Errors.hs [hunk 1]
```diff
@@ -261,6 +248,31 @@ data LabelType = TypeLabel | SkolemLabel String deriving (Show, Eq, Ord)
 -- | A map from rigid type variable name/unknown variable pairs to new variables.
 type UnknownMap = M.Map (LabelType, Unknown) Unknown
 
+-- |
+-- Extract nested error messages from wrapper errors
+--
+unwrapErrorMessage :: ErrorMessage -> Maybe ErrorMessage
+unwrapErrorMessage em = case em of
+  (ErrorCheckingKind _ err)       -> Just err
```
Comment (paf31) [inline src/Language/PureScript/Errors.hs]: This isn't going to work for certain errors. You need to recursively call `unwrapErrorMessage` on the wrapped error until you get down to an unwrapped error. On the plus side, this means you can get rid of the `Maybe`.

**Comment by mjgpy3** [comment]
@paf31: more like this? Do you know a way to ensure exhaustive matches in `errorCode` (or error in counter cases)? I was going to fail when the wrapped and unwrapped were equal, but I notice `ErrorMessage` isn't an instance of `Eq`. Is there a particular reason for that?

**Comment by paf31** [comment]
Looks great. For the exhaustivity issue, I would use a `case` at the top level, and then add a wildcard match with an `error` call inside. A better approach, allowing us to make use of GHC's exhaustivity checks, would be to extract the simple errors into another data type (`SimpleErrorMessage` maybe?)

**Comment by mjgpy3** [comment]
I'm getting a little confused, would it be enough to make `unwrapErrorMessage` exhaustive across `ErrorMessage` and have no wildcard matcher?

**Comment by mjgpy3** [comment]
I hope you don't mind all this back and forth. To be honest, I am fully enjoying the discussion.

**Comment by mjgpy3** [comment]
With the `SimpleErrorMessage` what type would it be, and how would it help the compiler detect non-exhaustive matching?

**Comment by paf31** [comment]
No problem at all. So `SimpleErrorMessage` would have all of the constructors which do not wrap another error, and `ErrorMessage` would have all the others, and an extra constructor for wrapping a `SimpleErrorMessage`. This would have the downside that all calls to `throwError` might have to be modified slightly. Then `unwrapErrorMessage` would have type `ErrorMessage -> SimpleErrorMessage`, and it would be exhaustive, checked by GHC.

**Comment by mjgpy3** [comment]
Excellent, thanks for the explanation. I see what I can do :+1:

**Comment by mjgpy3** [comment]
Hey @paf31, let me know your thoughts.

I wanted to make a function summarizing:
```
throwError . errorMessage $ SimpleErrorWrapper
```

but I couldn't get the type signature to work and it's getting late, so maybe another day :)

**Comment by paf31** [comment]
I think what you want is `throwError . errorMessage . SimpleErrorWrapper $ ...` but I would actually move the `SimpleErrorWrapper` constructor inside `errorMessage` since every call site seems to use it now.

Thank you for working on this - it's a really valuable change, since now we have the framework to give suggestions generally, which is something we really needed.

**Comment by mjgpy3** [comment]
Awesome, thank you!

**Comment by mjgpy3** [comment]
That's straing

**Comment by mjgpy3** [comment]
Oops

**Comment by mjgpy3** [comment]
@paf31, I pulled that constructor in. Let me know what you think

**Comment by paf31** [comment]
Looks great. Could you please do the following and then I'll get this merged into master:

- Split `errorMessage` into `errorMessage` (what it is now) and `singleError = MultipleErrors . pure` to replace the many instances of `MultipleErrors . (: [])`.
- Make sure the tests still pass
- Revert the files which didn't need to change (the uses of `$`), just to keep the history clean.
- Merge master up, and squash your commits into a single commit.

Also, could you please show me the output of, say `:type const 1 . const 2` in `psci`?

Thanks!

**Pull request closed without merging**